File: src/main/java/com/chatchatabc/jpademojava/application/rest/AuthController.java
package com.chatchatabc.jpademojava.application.rest;

import com.chatchatabc.jpademojava.application.dto.ErrorContent;
import com.chatchatabc.jpademojava.application.dto.user.UserLoginRequest;
import com.chatchatabc.jpademojava.application.dto.user.UserLoginResponse;
import com.chatchatabc.jpademojava.application.dto.user.UserRegisterRequest;
import com.chatchatabc.jpademojava.application.dto.user.UserRegisterResponse;
import com.chatchatabc.jpademojava.application.rest.jwt.JwtService;
import com.chatchatabc.jpademojava.domain.model.User;
import com.chatchatabc.jpademojava.domain.repository.UserRepository;
import com.chatchatabc.jpademojava.domain.service.UserService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.Optional;

@RestController
@RequestMapping("/api/auth")
public class AuthController {
    @Autowired
    private UserService userService;
    @Autowired
    private UserRepository userRepository;
    @Autowired
    private AuthenticationManager authenticationManager;
    @Autowired
    private JwtService jwtService;

    /**
     * User Login
     *
     * @param request
     * @return
     */
    @PostMapping("/login")
    public ResponseEntity<UserLoginResponse> login(
            @RequestBody UserLoginRequest request
    ) {
        try {
            Optional<User> queriedUser = userRepository.findByUsername(request.getUsername());
            if (queriedUser.isEmpty()) {
                throw new Exception("User not found");
            }

            // Authenticate User
            authenticationManager.authenticate(
                    new UsernamePasswordAuthenticationToken(
                            request.getUsername(),
                            request.getPassword()
                    )
            );

            // Generate JWT
            String token = jwtService.generateToken(queriedUser.get());
            HttpHeaders headers = new HttpHeaders();
            headers.set("X-Access-Token", token);
            return ResponseEntity.ok().headers(headers)
                    .body(new UserLoginResponse(queriedUser.get(), null));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                    .body(new UserLoginResponse(null, new ErrorContent("User Login Error", e.getMessage())));
        }
    }

    /**
     * User Registration
     *
     * @param request
     * @return
     */
    @PostMapping("/register")
    public ResponseEntity<UserRegisterResponse> register(
            @RequestBody UserRegisterRequest request
    ) {
        try {
            User newUser = userService.register(request);
            return ResponseEntity.status(HttpStatus.CREATED).body(new UserRegisterResponse(newUser, null));
        } catch (Exception e) {
            // TODO: Catch appropriate exception message
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                    .body(new UserRegisterResponse(null, new ErrorContent("User Register Error", e.getMessage())));
        }
    }

    /**
     * TODO: Check if username is available
     */

    /**
     * TODO: Check if email is available
     */
}

File: src/main/kotlin/com/chatchatabc/jpademo/application/rest/AuthController.kt
package com.chatchatabc.jpademo.application.rest

import com.chatchatabc.jpademo.application.dto.ErrorContent
import com.chatchatabc.jpademo.application.dto.user.UserLoginRequest
import com.chatchatabc.jpademo.application.dto.user.UserLoginResponse
import com.chatchatabc.jpademo.application.dto.user.UserRegisterRequest
import com.chatchatabc.jpademo.application.dto.user.UserRegisterResponse
import com.chatchatabc.jpademo.application.rest.jwt.JwtService
import com.chatchatabc.jpademo.domain.repository.UserRepository
import com.chatchatabc.jpademo.domain.service.UserService
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/auth")
class AuthController(
    private val userService: UserService,
    private val userRepository: UserRepository,
    private val authenticationManager: AuthenticationManager,
    private val jwtService: JwtService
) {
    /**
     * User Login
     */
    @PostMapping("/login")
    fun login(
        @RequestBody userLoginRequest: UserLoginRequest
    ): ResponseEntity<UserLoginResponse> {
        return try {
            val queriedUser = userRepository.findByUsername(userLoginRequest.username)
            if (queriedUser.isEmpty) {
                throw Exception("User not found")
            }

            // Authenticate user
            authenticationManager.authenticate(
                UsernamePasswordAuthenticationToken(
                    userLoginRequest.username,
                    userLoginRequest.password
                )
            )

            val token: String = jwtService.generateToken(queriedUser.get())
            val headers = HttpHeaders()
            headers.set("X-Access-Token", token)
            ResponseEntity.ok().headers(headers).body(UserLoginResponse(queriedUser.get(), null))
        } catch (e: Exception) {
            ResponseEntity.status(HttpStatus.BAD_REQUEST)
                .body(UserLoginResponse(null, ErrorContent("User Login Error", e.message)))
        }
    }

    /**
     * User Registration
     */
    @PostMapping("/register")
    fun register(
        @RequestBody userRegisterRequest: UserRegisterRequest
    ): ResponseEntity<UserRegisterResponse> {
        return try {
            val newUser = userService.register(userRegisterRequest)
            return ResponseEntity.status(HttpStatus.CREATED).body(UserRegisterResponse(newUser, null))
        } catch (e: Exception) {
            // TODO: Catch appropriate exception message
            ResponseEntity.badRequest()
                .body(UserRegisterResponse(null, ErrorContent("User Registration Error", e.message)))
        }
    }

    /**
     * TODO: Check if username is available
     */

    /**
     * TODO: Check if email is available
     */
}